File: scripts/database_integration_scripts_V2/legacy/57_add_entity_type_to_cities.py
"""57 — Add entity_type / entity_type_ids to cities.

Mirrors `enhance_db/src/bin/57_add_entity_type_to_cities.rs`.

  Inputs : data/all_humans/city_entity_types.json
              {qid: {types: [{id: Q..., label: ...}, ...]}}
  Output : cities.entity_type     (pipe-joined English labels)
           cities.entity_type_ids (pipe-joined Wikidata Q-ids)

Usage
-----
    python3 57_add_entity_type_to_cities.py
    python3 57_add_entity_type_to_cities.py --full
"""

from __future__ import annotations

import json
import sqlite3
import tempfile
from pathlib import Path

from tqdm import tqdm

from common import (
    ALL_HUMANS_DIR,
    add_column_if_missing,
    insert_rows,
    load_json,
    log,
    open_db,
    parse_run_mode,
)

JSON_PATH = ALL_HUMANS_DIR / "city_entity_types.json"


def run(conn: sqlite3.Connection, json_path: Path = JSON_PATH) -> int:
    log("[DB] 57: Add entity_type to cities...")
    types_map = load_json(json_path)
    log(f"[DB] {len(types_map)} cities with P31 data")

    add_column_if_missing(conn, "cities", "entity_type", "TEXT")
    add_column_if_missing(conn, "cities", "entity_type_ids", "TEXT")
    conn.execute("UPDATE cities SET entity_type = NULL, entity_type_ids = NULL")
    conn.commit()

    cur = conn.cursor()
    updated = 0
    for qid, val in tqdm(types_map.items(), desc="57", unit="city"):
        types = val.get("types") if isinstance(val, dict) else None
        if not types:
            continue
        labels: list[str] = []
        ids: list[str] = []
        for t in types:
            tid = (t.get("id") or "").strip()
            label = (t.get("label") or "").strip()
            if not tid:
                continue
            ids.append(tid)
            labels.append(label or tid)
        if not ids:
            continue
        cur.execute(
            "UPDATE cities SET entity_type = ?, entity_type_ids = ? WHERE id = ?",
            ("|".join(labels), "|".join(ids), qid),
        )
        updated += cur.rowcount
    conn.commit()
    log(f"[DB] updated {updated} cities")
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_cities_entity_type ON cities(entity_type)"
    )
    conn.commit()
    return updated


def _sample_main() -> None:
    fake = {
        "Q90": {"types": [
            {"id": "Q515", "label": "city"},
            {"id": "Q174844", "label": "megacity"},
        ]},
        "Q60": {"types": [{"id": "Q515", "label": "city"}]},
        "Q777": {"types": []},
    }
    with tempfile.TemporaryDirectory() as tmp:
        json_path = Path(tmp) / "city_entity_types.json"
        json_path.write_text(json.dumps(fake))
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db) as seed:
            seed.execute(
                "CREATE TABLE cities (id TEXT PRIMARY KEY, name_en TEXT)"
            )
            insert_rows(seed, "cities", [
                {"id": "Q90", "name_en": "Paris"},
                {"id": "Q60", "name_en": "New York"},
                {"id": "Q777", "name_en": "Mystery"},
            ])
        with open_db(db) as conn:
            run(conn, json_path=json_path)
            for r in conn.execute(
                "SELECT id, name_en, entity_type, entity_type_ids FROM cities"
            ):
                log(f"  {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()

File: scripts/database_integration_scripts/enhance_db/src/bin/57_add_entity_type_to_cities.rs
/// 57: Add `entity_type` and `entity_type_ids` columns to the cities table from
/// the Wikidata P31 (instance of) values previously fetched by
/// `extraction_scripts/all_humans/35_fetch_city_entity_types.py`.
///
/// Input : data/all_humans/city_entity_types.json
/// Output: cities.entity_type     (pipe-separated English labels,   e.g. "city|big city")
///         cities.entity_type_ids (pipe-separated Wikidata Q-ids,   e.g. "Q515|Q1549591")
use anyhow::{Context, Result};
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const TYPES_PATH: &str = "data/all_humans/city_entity_types.json";
const TASK_LOG: &str = "task.log";

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn column_exists(conn: &Connection, table: &str, col: &str) -> Result<bool> {
    let mut stmt = conn.prepare(&format!("PRAGMA table_info({})", table))?;
    let names: Vec<String> = stmt
        .query_map([], |row| row.get::<_, String>(1))?
        .filter_map(|r| r.ok())
        .collect();
    Ok(names.iter().any(|n| n == col))
}

fn main() -> Result<()> {
    log("=== Step 57: Add entity_type to cities ===");

    let types_json =
        fs::read_to_string(TYPES_PATH).with_context(|| format!("reading {}", TYPES_PATH))?;
    let types_map: HashMap<String, Value> =
        serde_json::from_str(&types_json).context("parsing city_entity_types.json")?;
    log(&format!(
        "[57] Loaded {} cities with P31 data",
        types_map.len()
    ));

    let mut conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    for (col, typ) in [("entity_type", "TEXT"), ("entity_type_ids", "TEXT")] {
        if column_exists(&conn, "cities", col)? {
            log(&format!("[57] cities.{} already exists — clearing values", col));
            conn.execute(&format!("UPDATE cities SET {} = NULL", col), [])?;
        } else {
            conn.execute_batch(&format!("ALTER TABLE cities ADD COLUMN {} {};", col, typ))?;
            log(&format!("[57] Added column cities.{}", col));
        }
    }

    let tx = conn.transaction()?;
    let updated_total: usize;
    {
        let mut stmt = tx.prepare(
            "UPDATE cities SET entity_type = ?1, entity_type_ids = ?2 WHERE id = ?3",
        )?;

        let pb = ProgressBar::new(types_map.len() as u64);
        pb.set_style(
            ProgressStyle::with_template(
                "  {msg} [{bar:40.cyan/blue}] {pos}/{len} ({percent}%) eta {eta}",
            )
            .unwrap()
            .progress_chars("=>-"),
        );
        pb.set_message("Updating cities");

        let mut updated = 0usize;
        for (qid, v) in &types_map {
            let types = v.get("types").and_then(|x| x.as_array());
            if types.is_none() {
                pb.inc(1);
                continue;
            }
            let arr = types.unwrap();

            let mut labels: Vec<String> = Vec::with_capacity(arr.len());
            let mut ids: Vec<String> = Vec::with_capacity(arr.len());
            for t in arr {
                let id = t.get("id").and_then(|x| x.as_str()).unwrap_or("").to_string();
                let label = t
                    .get("label")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string();
                if id.is_empty() {
                    continue;
                }
                ids.push(id);
                // fall back to id if label is empty — so the column is never silently lossy
                if label.is_empty() {
                    labels.push(ids.last().unwrap().clone());
                } else {
                    labels.push(label);
                }
            }

            if ids.is_empty() {
                pb.inc(1);
                continue;
            }

            let labels_joined = labels.join("|");
            let ids_joined = ids.join("|");
            let rows = stmt.execute(params![labels_joined, ids_joined, qid])?;
            updated += rows as usize;
            pb.inc(1);
        }
        pb.finish_with_message("Updating cities done");
        updated_total = updated;
    }
    tx.commit()?;
    log(&format!("[57] Updated {} city rows", updated_total));

    // Indexes for common filtering
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_cities_entity_type ON cities(entity_type);",
    )?;

    // Summary / sanity checks
    let total: i64 = conn.query_row("SELECT COUNT(*) FROM cities", [], |r| r.get(0))?;
    let with_type: i64 = conn.query_row(
        "SELECT COUNT(*) FROM cities WHERE entity_type IS NOT NULL",
        [],
        |r| r.get(0),
    )?;
    log(&format!(
        "[57] cities total={} with_entity_type={} coverage={:.1}%",
        total,
        with_type,
        100.0 * with_type as f64 / total.max(1) as f64
    ));

    // Top 20 entity_type strings by frequency
    log("[57] Top 20 entity_type values:");
    let mut stmt = conn.prepare(
        "SELECT entity_type, COUNT(*) c FROM cities
         WHERE entity_type IS NOT NULL
         GROUP BY entity_type ORDER BY c DESC LIMIT 20",
    )?;
    let rows: Vec<(String, i64)> = stmt
        .query_map([], |r| Ok((r.get::<_, String>(0)?, r.get::<_, i64>(1)?)))?
        .filter_map(|r| r.ok())
        .collect();
    for (et, c) in rows {
        let truncated: String = et.chars().take(120).collect();
        log(&format!("[57]   {:>8}  {}", c, truncated));
    }

    // Quick look at the examples we started with
    let mut stmt = conn.prepare(
        "SELECT id, name_en, entity_type FROM cities
         WHERE id IN ('Q100','Q30','Q771','Q60','Q90','Q1537')",
    )?;
    let rows: Vec<(String, Option<String>, Option<String>)> = stmt
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)))?
        .filter_map(|r| r.ok())
        .collect();
    log("[57] Spot check:");
    for (id, name, et) in rows {
        log(&format!(
            "[57]   {} {} -> {}",
            id,
            name.as_deref().unwrap_or("?"),
            et.as_deref().unwrap_or("NULL")
        ));
    }

    log("=== Step 57 complete ===");
    Ok(())
}
